### DISCUSSION TITLE
Augmentation with bounding boxes.

### DISCUSSION THREAD

**Discussion by bot66** [kornia/kornia#2399] [Potential Issue] [answered]
When using `RandomAffine(degrees=30,translate=None,scale=None,shear=None,p=0.5)` to rotate an image with bounding boxes,  bounding boxes are not tightly aligned with the object, is this a correct behavior? (see examples below) 

**Example 1:**
input
![raw_16](https://github.com/kornia/kornia/assets/26666338/894b29ca-1dd4-4361-a4d9-434356b7039f)

output
![16](https://github.com/kornia/kornia/assets/26666338/5bc0243c-4210-4f9f-9d93-66e29728da74)

**Example 2:**
input
![raw_5](https://github.com/kornia/kornia/assets/26666338/ced10d67-6a94-45ed-ba5a-7ab08554086e)

output
![5](https://github.com/kornia/kornia/assets/26666338/34df72ed-0254-41f8-b825-66692c1562b0)

When using shear `RandomAffine(degrees=0,translate=None,scale=None,shear=15,p=0.5)`, it has the similar behavior:

input

![raw_17](https://github.com/kornia/kornia/assets/26666338/30f1362f-2441-4486-8d0b-8da4e48c6fb5)

output

![17](https://github.com/kornia/kornia/assets/26666338/8ff41f4d-e342-4f7a-aa9e-029d51993f88)

**Comment by edgarriba** [comment]
@bot66 do you have code to reproduce that ?

**Reply by bot66** [reply]
Yes, I will provide the code later, and I also found the affine translation for bbox also not working well.
When using `K.RandomAffine(degrees=0,translate=(0.1,0.1),scale=None,shear=None,p=0.5)`
input:
![raw_9](https://github.com/kornia/kornia/assets/26666338/e91f36d2-e890-4bb4-af91-79d56aa11684)
output:
![9](https://github.com/kornia/kornia/assets/26666338/80914f5e-4421-4571-8819-6d15f2aa9200)

**Comment by bot66** [comment]
Hi @edgarriba This is my code , input data and result.


```python
import kornia.augmentation as K
import torch
from torchvision.utils import draw_bounding_boxes,save_image
import cv2
import numpy as np
from torchvision.transforms import ToTensor


def xywhn2xyxy(x, w=640, h=640, padw=0, padh=0):
    # Convert nx4 boxes from [x, y, w, h] normalized to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
    y = x.clone() if isinstance(x, torch.Tensor) else np.copy(x)
    y[..., 0] = w * (x[..., 0] - x[..., 2] / 2) + padw  # top left x
    y[..., 1] = h * (x[..., 1] - x[..., 3] / 2) + padh  # top left y
    y[..., 2] = w * (x[..., 0] + x[..., 2] / 2) + padw  # bottom right x
    y[..., 3] = h * (x[..., 1] + x[..., 3] / 2) + padh  # bottom right y
    return y

def letterbox(im, new_shape=(640, 640), color=(114, 114, 114), auto=False, scaleFill=False, scaleup=True, stride=32):
    # Resize and pad image while meeting stride-multiple constraints
    shape = im.shape[:2]  # current shape [height, width]
    if isinstance(new_shape, int):
        new_shape = (new_shape, new_shape)

    # Scale ratio (new / old)
    r = min(new_shape[0] / shape[0], new_shape[1] / shape[1])
    if not scaleup:  # only scale down, do not scale up (for better val mAP)
        r = min(r, 1.0)

    # Compute padding
    ratio = r, r  # width, height ratios
    new_unpad = int(round(shape[1] * r)), int(round(shape[0] * r))
    dw, dh = new_shape[1] - new_unpad[0], new_shape[0] - new_unpad[1]  # wh padding
    if auto:  # minimum rectangle
        dw, dh = np.mod(dw, stride), np.mod(dh, stride)  # wh padding
    elif scaleFill:  # stretch
        dw, dh = 0.0, 0.0
        new_unpad = (new_shape[1], new_shape[0])
        ratio = new_shape[1] / shape[1], new_shape[0] / shape[0]  # width, height ratios

    dw /= 2  # divide padding into 2 sides
    dh /= 2

    if shape[::-1] != new_unpad:  # resize
        im = cv2.resize(im, new_unpad, interpolation=cv2.INTER_LINEAR)
    top, bottom = int(round(dh - 0.1)), int(round(dh + 0.1))
    left, right = int(round(dw - 0.1)), int(round(dw + 0.1))
    im = cv2.copyMakeBorder(im, top, bottom, left, right, cv2.BORDER_CONSTANT, value=color)  # add border
    return im, ratio, (dw, dh)


if __name__=="__main__":
    w,h=640,480
    image_path="2008_003985.jpg"
    label_path="2008_003985.txt"
    
    img=cv2.imread(image_path)

    #resize image
    img, ratio, pad=letterbox(img,(h,w))

    #load yolo label
    with open(label_path, 'r') as f:
        lb = [x.split() for x in f.read().strip().splitlines() if len(x)]
        lb = np.array(lb, dtype=np.float32)
    
    #make tensors
    image_tensor=ToTensor()(img[:,:,::-1].copy()) #bgr->rgb->tensor
    lb[:, 1:]= xywhn2xyxy(lb[:, 1:], w, h, padw=pad[0], padh=pad[1])
    label_tensor=torch.from_numpy(lb) #cls_idx,xyxy
    

    augs=[
        K.RandomAffine(degrees=0,translate=(0.1,0.1),scale=None,shear=None,p=1),# translate
        # K.RandomAffine(degrees=30,translate=None,scale=None,shear=None,p=1),# rotate
        # K.RandomAffine(degrees=0,translate=None,scale=None,shear=15,p=1),# shear
    ]

    augmentor=K.AugmentationSequential(*augs,
                                    data_keys=["input","bbox_xyxy"], 
                                    same_on_batch=False,
                                    keepdim=False,
                                    random_apply=(1,6))
    
    # apply augmentation
    new_label_tensor=label_tensor.clone()
    new_image_tensor,new_label_tensor[:,1:]=augmentor(image_tensor,new_label_tensor[:,1:])

    #draw bounding box
    voc_labelmap={0: "aeroplane",
        1: "bicycle",
        2: "bird",
        3: "boat",
        4: "bottle",
        5: "bus",
        6: "car",
        7: "cat",
        8: "chair",
        9: "cow",
        10: "diningtable",
        11: "dog",
        12: "horse",
        13: "motorbike",
        14: "person",
        15: "pottedplant",
        16: "sheep",
        17: "sofa",
        18: "train",
        19: "tvmonitor"
    }
    
    # draw image before augmentation
    img_tensor=image_tensor.mul(255).add(0.5).clamp(0,255).to(torch.uint8)
    h,w=img_tensor.shape[1:]
    cls_idx,bbox=label_tensor[...,0],label_tensor[...,1:]
    labels=[voc_labelmap[int(idx.item())] for idx in cls_idx]
    img_tensor=draw_bounding_boxes(img_tensor,bbox,labels,colors='green',width=5) # cpu tensor
    save_image(img_tensor/255,f"before_augmented_{image_path}")


    # draw image after augmentation
    img_tensor=new_image_tensor[0].mul(255).add(0.5).clamp(0,255).to(torch.uint8)
    h,w=img_tensor.shape[1:]
    cls_idx,bbox=new_label_tensor[...,0],new_label_tensor[...,1:]
    labels=[voc_labelmap[int(idx.item())] for idx in cls_idx]
    img_tensor=draw_bounding_boxes(img_tensor,bbox,labels,colors='green',width=5) # cpu tensor
    save_image(img_tensor/255,f"after_augmented_{image_path}")


```


input data:
![2008_003985](https://github.com/kornia/kornia/assets/26666338/f081e166-e404-4c8f-bcb9-1a4730a1b4c6)
[2008_003985.txt](https://github.com/kornia/kornia/files/11704725/2008_003985.txt)

input:
![before_augmented_2008_003985](https://github.com/kornia/kornia/assets/26666338/afbda369-a4dd-410a-b2a6-bda55ae0b9ef)
output:
![after_augmented_2008_003985](https://github.com/kornia/kornia/assets/26666338/e5bbfa34-130f-41ed-bb00-7bc107111045)

**Reply by bot66** [reply]
@edgarriba It sometimes works, sometimes not, kinda random.

**Comment by bot66** [comment, marked as answer]
Hi @edgarriba, I found the problem, It's because when using `AugmentationSequential`,  If set the `random_apply(1,6)`, It will translate the image multiple times, and some image information is already lost during those transforms, but the bounding boxes are not clipped.

For example, when translating left(-5) and then translating right(+5), the bounding box retains the original, but the image is not the original, like this:

![after_augmented_seq_0_2008_003985](https://github.com/kornia/kornia/assets/26666338/23a96602-3106-47a3-9b35-8c2c44fca183)

~~Is it possible that clip the bounding box after every time it is transformed in the latest version of kornia ?~~

~~Related issue:https://github.com/kornia/kornia/issues/1142#issuecomment-912943068~~

Ok, I understand, I will use `RandomAffine` outside of the `AugmentationSequential`.

**Reply by edgarriba** [reply]
@bot66 in any case i still think there's a bug somewhere ? @johnnv1 @shijianjian

**Reply by johnnv1** [reply]
I think I've seen the same issue in one of the tutorials -- I believe it's a bug, but I never take a better look
